presc/spatial_distribution/categorical_distance.py

```diff
@@ -0,0 +1,509 @@
+
+    def overlap(self, dpoint1, dpoint2):
+        """Computes the overlap similarity measure for categorical data"""
+
+        if list(dpoint1.index) == list(dpoint2.index):
+            overlap_score = len(set(dpoint1) & set(dpoint2)) / len(dpoint1)
+            return overlap_score
+        else:
+            # write exception
+
+            pass
+
+    def lin(self, dpoint1, dpoint2):
+        d1_attributes = list(dpoint1.index)
+        d2_attributes = list(dpoint2.index)
+
+        if d1_attributes == d2_attributes:
+            lin = 0
+            weight = self.lin_weight(dpoint1, dpoint2)
+            for attribute in d1_attributes:
+                frequency_dpoint1 = self.empirical_frequency(
+                    dpoint1[attribute], attribute
+                )
+                frequency_dpoint2 = self.empirical_frequency(
+                    dpoint2[attribute], attribute
+                )
+                if dpoint1[attribute] == dpoint2[attribute]:
+                    lin = lin + 2 * np.log(frequency_dpoint1)
+                else:
+                    lin = lin + 2 * np.log(frequency_dpoint1 + frequency_dpoint2)
+        return weight * lin
+
+    def lin_weight(self, dpoint1, dpoint2):
+        d1_attributes = list(dpoint1.index)
+        weight_denominator = 0
+        for attribute in d1_attributes:
+            frequency_dpoint1 = self.empirical_frequency(dpoint1[attribute], attribute)
+            frequency_dpoint2 = self.empirical_frequency(dpoint2[attribute], attribute)
+            weight_denominator = (
+                weight_denominator
+                + np.log(frequency_dpoint1)
+                + np.log(frequency_dpoint2)
+            )
+
+        return 1 / weight_denominator
+
+        return
+
+    def empirical_frequency(self, value, attribute):
+        counts = self._counts_per_attribute[attribute][value]
+        return counts / self._data_len
+
+    def goodall_frequency(self, counts):
+        """Computes a probability estimate of an attribute required to compute
+        goodall similarity measures"""
+        estimated_freq = (counts * (counts - 1)) / (
+            (self._data_len) * (self._data_len - 1)
+        )
+        return estimated_freq
+
+    def distance_to_data(self, dpoint, metric, distance_sample=0.013):
+        """Computes an estimate of the average distance of a data point to every
+        other data points by taking a random sample of a given size and avereging
+        the result
+        Args:
+             dpoint: data point to work with
+             metric: string, similarity metric used to compute the distance
+             distance_sample: float for 0 to 1 that indicates the percentage
+                              from the total data that should be used to
+                              estimate the average distance"""
+
+        if distance_sample < 1:
+            distance_sample = round(distance_sample * self._data_len)
+        elif distance_sample > self._data_len:
```

Comment:
the user can input a number of sample points the function can use. I originally forgot to add this to the documentation. In case the number is greater than 1 the program will use it as the exact number of points used to estimate the average distance